src/segmentedfilesystem/PacketStructure.java

```diff
@@ -0,0 +1,45 @@
+package segmentedfilesystem;
+import java.net.DatagramPacket;
+import java.util.Arrays;
+
+public class PacketStructure {
+    private byte status;
+    private byte ID;
+    private byte[] pnumber = new byte[2];
+    private byte[] data;
+    private int length;
+
+    PacketManager manager = new PacketManager();
+    public PacketStructure(DatagramPacket packet){
+        status = packet.getData()[0];
+        ID = packet.getData()[1];
+        length = packet.getLength();
+        if(determineType(packet)) {
+            data = packet.getData();
+        }
+        else{
+            pnumber[0] = packet.getData()[2];
+            pnumber[1] = packet.getData()[3];
+            data = Arrays.copyOfRange(packet.getData(),4,packet.getData().length-4);

```

Comment:
This is why you were getting the extra data at the end. You were using `packet.getData().length` (which returns the size of the array, which is always 1028) instead of `packet.getLength()`, which is the number of _meaningful_ bytes in that packet. That mistake meant your final packet always had a bunch of random stuff in it.